`ImmutableIndexedObjectListUsingArray` handles null array incorrectly

`ImmutableIndexedObjectListUsingArray` allows its `dataArray` to be null. However, some of its methods like `withAddedObject` or `stream` are handling this case incorrectly. 

This issue ultimately causes flakiness of `RandomVirtualMapReconnectTests#randomOperationsReconnectTest` by cascade of errors: 

```
13:44:29.035   6792 <merkledb: Compacting #3> [IdToHashChunk_compact_0_0] Compaction failed
java.lang.NullPointerException: Cannot read the array length because "array" is null
	at java.base/java.util.Arrays.stream(Arrays.java:5464)
	at com.swirlds.merkledb.collections.ImmutableIndexedObjectListUsingArray.stream(ImmutableIndexedObjectListUsingArray.java:145)
	at com.swirlds.merkledb.files.DataFileCollection.getAllCompletedFiles(DataFileCollection.java:297)
	at com.swirlds.merkledb.files.DataFileCompactor.compactSingleLevel(DataFileCompactor.java:224)
	at com.swirlds.merkledb.MerkleDbCompactionCoordinator$CompactionTask.call(MerkleDbCompactionCoordinator.java:629)
	at com.swirlds.merkledb.MerkleDbCompactionCoordinator$CompactionTask.call(MerkleDbCompactionCoordinator.java:568)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:328)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
	at org.hiero.consensus.concurrent.framework.internal.AbstractThreadConfiguration.lambda$wrapRunnableWithSnapshot$0(AbstractThreadConfiguration.java:654)
	at java.base/java.lang.Thread.run(Thread.java:1474)
```
Then
```
13:44:29.124   6881 <virtual-pipeline: lifecycle #0> [state] Interrupted while waiting on executors to shutdown
java.lang.InterruptedException
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:1789)
	at java.base/java.util.concurrent.ThreadPoolExecutor.awaitTermination(ThreadPoolExecutor.java:1421)
	at java.base/java.util.concurrent.Executors$DelegatedExecutorService.awaitTermination(Executors.java:691)
	at com.swirlds.merkledb.MerkleDbDataSource.shutdownThreadsAndWait(MerkleDbDataSource.java:1187)
	at com.swirlds.merkledb.MerkleDbDataSource.close(MerkleDbDataSource.java:939)
	at com.swirlds.virtualmap.datasource.VirtualDataSource.close(VirtualDataSource.java:39)
	at com.swirlds.virtualmap.VirtualMap.closeDataSource(VirtualMap.java:865)
	at com.swirlds.virtualmap.VirtualMap.onShutdown(VirtualMap.java:857)
	at com.swirlds.virtualmap.internal.pipeline.VirtualPipeline.fireOnShutdown(VirtualPipeline.java:591)
	at com.swirlds.virtualmap.internal.pipeline.VirtualPipeline.shutdown(VirtualPipeline.java:524)
	at com.swirlds.virtualmap.internal.pipeline.VirtualPipeline.doWork(VirtualPipeline.java:507)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:545)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:328)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
	at org.hiero.consensus.concurrent.framework.internal.AbstractThreadConfiguration.lambda$wrapRunnableWithSnapshot$0(AbstractThreadConfiguration.
``` 
Then

```
13:44:29.131   6888 <virtual-pipeline: lifecycle #0> Could not close the dataSource after all copies were destroyed
java.io.IOException: Interrupted while waiting for shutdown to finish.
	at com.swirlds.merkledb.MerkleDbDataSource.shutdownThreadsAndWait(MerkleDbDataSource.java:1196)
	at com.swirlds.merkledb.MerkleDbDataSource.close(MerkleDbDataSource.java:939)
	at com.swirlds.virtualmap.datasource.VirtualDataSource.close(VirtualDataSource.java:39)
	at com.swirlds.virtualmap.VirtualMap.closeDataSource(VirtualMap.java:865)
	at com.swirlds.virtualmap.VirtualMap.onShutdown(VirtualMap.java:857)
	at com.swirlds.virtualmap.internal.pipeline.VirtualPipeline.fireOnShutdown(VirtualPipeline.java:591)
	at com.swirlds.virtualmap.internal.pipeline.VirtualPipeline.shutdown(VirtualPipeline.java:524)
	at com.swirlds.virtualmap.internal.pipeline.VirtualPipeline.doWork(VirtualPipeline.java:507)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:545)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:328)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
	at org.hiero.consensus.concurrent.framework.internal.AbstractThreadConfiguration.lambda$wrapRunnableWithSnapshot$0(AbstractThreadConfiguration.java:654)
	at java.base/java.lang.Thread.run(Thread.java:1474)
Caused by: java.lang.InterruptedException
```

and finally 
```
java.util.concurrent.RejectedExecutionException: Task com.swirlds.virtualmap.internal.pipeline.VirtualPipeline$$Lambda/0x00000000873cd6e0@7e72a6cc rejected from java.util.concurrent.ThreadPoolExecutor@4098dd77[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 1]

	at java.base/java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2032)
	at java.base/java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:787)
	at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1328)
	at java.base/java.util.concurrent.Executors$DelegatedExecutorService.execute(Executors.java:665)
	at com.swirlds.virtualmap.internal.pipeline.VirtualPipeline.pausePipelineAndExecute(VirtualPipeline.java:547)
	at com.swirlds.virtualmap.internal.pipeline.VirtualPipeline.pausePipelineAndRun(VirtualPipeline.java:358)
	at com.swirlds.virtualmap.VirtualMap.detach(VirtualMap.java:1218)
	at com.swirlds.virtualmap.internal.reconnect.TeacherPullVirtualTreeView.<init>(TeacherPullVirtualTreeView.java:54)
	at com.swirlds.virtualmap.VirtualMap.buildTeacherView(VirtualMap.java:1262)
	at com.swirlds.common.test.fixtures.merkle.util.MerkleTestUtils.testSynchronization(MerkleTestUtils.java:164)
	at com.swirlds.common.test.fixtures.merkle.util.MerkleTestUtils.hashAndTestSynchronization(MerkleTestUtils.java:310)
	at com.swirlds.platform.reconnect.RandomVirtualMapReconnectTests.randomOperationsReconnectTest(RandomVirtualMapReconnectTests.java:226)
[skipped]
	Suppressed: org.opentest4j.AssertionFailedError: Expected 0 open databases.
```

--- a/platform-sdk/swirlds-merkledb/src/main/java/com/swirlds/merkledb/collections/ImmutableIndexedObjectListUsingArray.java
+++ b/platform-sdk/swirlds-merkledb/src/main/java/com/swirlds/merkledb/collections/ImmutableIndexedObjectListUsingArray.java
@@ -87,6 +87,10 @@ public class ImmutableIndexedObjectListUsingArray<T extends IndexedObject> imple
             return this;
         }
 
+        if (isEmpty()) {
+            return new ImmutableIndexedObjectListUsingArray<>(arrayProvider, List.of(newObject));
+        }
+
         // Create a temp array list with just non-null objects that belong to the new list
         final List<T> newDataArray = new ArrayList<>(Arrays.asList(dataArray));
         newDataArray.removeIf(next -> next == null || next.getIndex() == newObject.getIndex());
@@ -142,7 +146,7 @@ public class ImmutableIndexedObjectListUsingArray<T extends IndexedObject> imple
     /** {@inheritDoc} */
     @Override
     public Stream<T> stream() {
-        return Arrays.stream(dataArray).filter(Objects::nonNull);
+        return isEmpty() ? Stream.empty() : Arrays.stream(dataArray).filter(Objects::nonNull);
     }
 
     @Override

--- a/platform-sdk/swirlds-merkledb/src/test/java/com/swirlds/merkledb/collections/ImmutableIndexedObjectListUsingArrayTest.java
+++ b/platform-sdk/swirlds-merkledb/src/test/java/com/swirlds/merkledb/collections/ImmutableIndexedObjectListUsingArrayTest.java
@@ -146,6 +146,29 @@ class ImmutableIndexedObjectListUsingArrayTest {
         assertNull(emptyList.get(42), "Empty lists should always return null");
     }
 
+    @Test
+    @Order(10)
+    void emptyListProducesEmptyStream() {
+        final ImmutableIndexedObjectList<TestIndexedObject> emptyList =
+                factoryForReaderToTest(new TestIndexedObject[0]);
+
+        assertEquals(0, emptyList.stream().count(), "Empty lists should produce empty streams");
+    }
+
+    @Test
+    @Order(11)
+    void withAddedObjectOnEmptyListAddsObject() {
+        final ImmutableIndexedObjectList<TestIndexedObject> emptyList =
+                factoryForReaderToTest(new TestIndexedObject[0]);
+        final TestIndexedObject object42 = new TestIndexedObject(42);
+
+        final ImmutableIndexedObjectList<TestIndexedObject> updatedList = emptyList.withAddedObject(object42);
+
+        assertEquals(1, updatedList.size(), "Adding to an empty list should create a singleton list");
+        assertSame(object42, updatedList.get(42), "Added object should be retrievable by index");
+        assertSame(object42, updatedList.getLast(), "Added object should become the last element");
+    }
+
     private static void checkRange(int max) {
         max++; // add one to make range exclusive from inclusive
         TestIndexedObject[] objects = list.stream().toArray(TestIndexedObject[]::new);

